horovod/spark/torch/estimator.py

```diff
@@ -0,0 +1,434 @@
+# Copyright 2019 Uber Technologies, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+from __future__ import absolute_import
+
+import horovod.spark.common._namedtuple_fix
+
+import io
+import numbers
+import time
+
+from horovod.run.common.util import codec
+from pyspark import keyword_only
+from pyspark.ml import Estimator, Model
+from pyspark.ml.param.shared import Param, Params
+from pyspark.ml.util import MLWritable, MLReadable
+
+from horovod.spark.common import util
+from horovod.spark.common.backend import SparkBackend
+from horovod.spark.common.params import EstimatorParams, ModelParams
+from horovod.spark.common.serialization import \
+    HorovodParamsWriter, HorovodParamsReader
+from horovod.spark.torch import remote
+from horovod.spark.torch.util import deserialize_fn, serialize_fn, \
+    save_into_bio
+
+import torch
+import torch.utils.data
+
+
+def _torch_param_serialize(param_name, param_val):
+    if param_name in [EstimatorParams.backend.name, EstimatorParams.store.name]:
+        # We do not serialize backend and store. These params have to be regenerated for each
+        # run of the pipeline
+        return None
+
+    if param_val is None:
+        return None
+
+    return codec.dumps_base64(param_val)
+
+
+class TorchEstimatorParamsWriter(HorovodParamsWriter):
+    def saveImpl(self, path):
+        # Write the parameters
+        HorovodParamsWriter.saveMetadata(self.instance, path, self.sc,
+                                         param_serializer_fn=_torch_param_serialize)
+
+
+class TorchEstimatorParamsWritable(MLWritable):
+    def write(self):
+        return TorchEstimatorParamsWriter(self)
+
+
+class TorchEstimatorParamsReader(HorovodParamsReader):
+    def _deserialize_dict(self, dict_values):
+        deserialized_dict = dict()
+        for key, val in dict_values.items():
+            if val is None:
+                deserialized_dict[key] = None
+            else:
+                deserialized_dict[key] = codec.loads_base64(val)
+        return deserialized_dict
+
+
+class TorchEstimatorParamsReadable(MLReadable):
+    @classmethod
+    def read(cls):
+        """Returns a DefaultParamsReader instance for this class."""
+        return TorchEstimatorParamsReader(cls)
+
+
+class TorchEstimator(Estimator, EstimatorParams, TorchEstimatorParamsWritable,
```

Comment:
has 4 methods in common with `KerasEstimator`: `_get_or_create_backend`, `fit_on_parquet`, `_fit_on_parquet`, `_fit`, `_has_checkpoint`. Maybe worth moving them into common base class.

Reply:
Good point.  Abstracted into `HorovodEstimator` in `horovod.spark.common.estimator`.